examples/virtio-snd/Makefile

```diff
@@ -1,311 +1,60 @@
-							--etc $(SND_DRIVER_VM_ETC)
-	
-$(CLIENT_VM_INITRD_OVERLAYED): $(addprefix $(BUILD_DIR)/, $(CLIENT_VM_USERLEVEL_ELFS)) \
-								$(addprefix $(BUILD_DIR)/, $(CLIENT_VM_USERLEVEL_INIT))
-	$(LIBVMM_TOOLS)/packrootfs $(CLIENT_VM_INITRD) $(CLIENT_VM_ROOTFS_TMP_DIR) -o $(CLIENT_VM_INITRD_OVERLAYED) \
-							--home $(addprefix $(BUILD_DIR)/, $(CLIENT_VM_USERLEVEL_ELFS))
-
-
-$(SND_DRIVER_VM_DTB): $(SND_DRIVER_VM_BASE_DTS) $(SND_DRIVER_VM_DTS_OVERLAYS)
-	sh ../../tools/dtscat $^ > $(BUILD_DIR)/snd_driver_vm.dts
-	$(DTC) -q -I dts -O dtb $(BUILD_DIR)/snd_driver_vm.dts > $@
-
-$(BUILD_DIR)/snd_driver_images.o: $(LIBVMM_TOOLS)/package_guest_images.S $(SND_DRIVER_VM_LINUX) $(SND_DRIVER_VM_DTB) $(SND_DRIVER_VM_INITRD_OVERLAYED)
-	$(CC) -c -g3 -x assembler-with-cpp \
-					-DGUEST_KERNEL_IMAGE_PATH=\"$(SND_DRIVER_VM_LINUX)\" \
-					-DGUEST_DTB_IMAGE_PATH=\"$(SND_DRIVER_VM_DTB)\" \
-					-DGUEST_INITRD_IMAGE_PATH=\"$(SND_DRIVER_VM_INITRD_OVERLAYED)\" \
-					-target aarch64-none-elf \
-					$< -o $@
-
-$(CLIENT_VM_DTB): $(CLIENT_VM_BASE_DTS) $(CLIENT_VM_DTS_OVERLAYS)
-	sh ../../tools/dtscat $^ > $(BUILD_DIR)/client_vm.dts
-	$(DTC) -q -I dts -O dtb $(BUILD_DIR)/client_vm.dts > $@
-
-$(BUILD_DIR)/client_images.o: $(LIBVMM_TOOLS)/package_guest_images.S $(CLIENT_VM_LINUX) $(CLIENT_VM_DTB) $(CLIENT_VM_INITRD_OVERLAYED)
-	$(CC) -c -g3 -x assembler-with-cpp \
-					-DGUEST_KERNEL_IMAGE_PATH=\"$(CLIENT_VM_LINUX)\" \
-					-DGUEST_DTB_IMAGE_PATH=\"$(CLIENT_VM_DTB)\" \
-					-DGUEST_INITRD_IMAGE_PATH=\"$(CLIENT_VM_INITRD_OVERLAYED)\" \
-					-target aarch64-none-elf \
-					$< -o $@
-
-$(BUILD_DIR)/user_sound/%.o: $(UIO_DRIVERS_DIR)/snd/%.c
-	mkdir -p $(BUILD_DIR)/user_sound
-	$(CC_USERLEVEL) -c $(CFLAGS_USERLEVEL) $^ -o $@
-
-$(BUILD_DIR)/user_sound.elf: $(BUILD_DIR)/user_sound/main.o $(BUILD_DIR)/user_sound/stream.o $(BUILD_DIR)/user_sound/queue.o $(BUILD_DIR)/user_sound/convert.o
-	$(CC_USERLEVEL) $(CFLAGS_USERLEVEL) $^ -o $@
-	patchelf --set-interpreter /lib64/ld-linux-aarch64.so.1 $@
-
-$(BUILD_DIR)/%.elf: $(VM_USERLEVEL_DIR)/%.c
-	$(CC_USERLEVEL) $(CFLAGS_USERLEVEL) $^ -o $@
-	patchelf --set-interpreter /lib64/ld-linux-aarch64.so.1 $@
-
-$(BUILD_DIR)/%.o: %.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/util/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/arch/aarch64/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/arch/aarch64/vgic/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/virtio/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
+export BUILD_DIR:=$(abspath $(BUILD_DIR))
+export MICROKIT_SDK:=$(abspath $(MICROKIT_SDK))
+export VIRTIO_EXAMPLE:=$(abspath .)
 
-$(BUILD_DIR)/sound_%.o: $(SDDF_SOUND_COMPONENTS)/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
+export TARGET := aarch64-none-elf
+export CC := clang
+export CC_USERLEVEL := zig cc
```

Comment:
Shouldn't require zig to build

Reply:
Currently it cannot be compiled with gcc or clang due to cross compilation issues relating to Nix and `libalsa`. I agree that this should be changed but I think it should be done in another PR.

Reply:
Problem is this prevents compilation on systems without zig.  Is there any reason why the user-level compiler cannot be the same compiler as for the rest of the system?

Reply:
Zig is the only one that currently works when using Nix which is what I have to use on macOS. I agree that this needs to be fixed but this is an existing issue from the existing build system (not a regression made by this PR). I think it should be fixed in another PR so this PR doesn't get blocked

Reply:
You obviously need to have enough of a cross-compilation environment installed.  This needs documenting somewhere.

On Debian, I can use aarch64-linux-gnu-gcc providing I've installed the aarch64 version of libasound2-dev
or clang with --target=aarch64-linux-gnu

Reply:
Since ALSA is a dynamic library we have to use a glibc based toolchain which `clang` is not.

So we'd either have to change the example to use `aarch64-linux-gnu-` everywhere, or something like `zig` which can compile to glibc.

Reply:
And also, like Alex said, when you want to get all of this working on macOS instead of Linux it becomes very complicated.

Reply:
Hmm, `clang -target aarch64-linux-gnu` does work for me. The question is whether it works on macOS now.

Reply:
I can't remember exactly but clang and gcc fail for different reasons. One of them tries to use the macOS standard library and gets confused and the other fails with error messages about `-iframework` (a macOS thing).

@erichchan999 did try and/or have issues compiling your userlevel block drivers with clang / gcc?

Reply:
Looks like `clang` does not package stdlib headers for other OSes, unlike Zig.

This means we'd either have to switch to `aarch64-linux-gnu-gcc` for everything, or use `zig`.

Reply:
If I recall correctly I only had success using gcc/zig for compiling my userlevel programs. I tried using clang before but ran into some issues trying to pass in -D_GNU_SOURCE or something along those lines for the block uio driver, which is separate problem to this whole discussion thread.

Reply:
Offline: going to fix this in another PR since I am only here for another 3 weeks

Reply:
This really needs access to the libraries and headers in the buildroot that generated the root FS, or build all tools 'static' so they're independent of the environment.